// @progress/kendo-angular-inputs/dist/npm/common/dom-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var UNTOUCHED = 'ng-untouched';
var toClassList = function (classNames) { return String(classNames).trim().split(' '); };
/**
 * @hidden
 */
exports.hasClass = function (element, className) {
    return Boolean(toClassList(element.className).find(function (name) { return name === className; }));
};
/**
 * @hidden
 */
function invokeElementMethod(element, name) {
    var args = [];
    for (var _i = 2; _i < arguments.length; _i++) {
        args[_i - 2] = arguments[_i];
    }
    if (element && element.nativeElement) {
        return element.nativeElement[name].apply(element.nativeElement, args);
    }
}
exports.invokeElementMethod = invokeElementMethod;
/**
 * @hidden
 */
function isDocumentAvailable() {
    return typeof document !== 'undefined';
}
exports.isDocumentAvailable = isDocumentAvailable;
/**
 * @hidden
 */
exports.isUntouched = function (element) {
    return element && element.nativeElement && exports.hasClass(element.nativeElement, UNTOUCHED);
};
/**
 * @hidden
 */
exports.containsFocus = function (hostElement, contender) {
    return hostElement && contender && (hostElement === contender || hostElement.contains(contender));
};
var w = function () { return Math.floor((1 + Math.random()) * 0x10000).toString(16).substring(1); };
/**
 * @hidden
 */
exports.guid = function () { return "" + w() + w() + "-" + w() + "-" + w() + "-" + w() + "-" + w() + w() + w(); };
